pkg/controller/master/upgradelog/controller.go

```diff
@@ -0,0 +1,518 @@
+	upgradeLogFluentdAnnotation       = "harvesterhci.io/fluentd"
+	upgradeLogFluentBitReady          = "FluentBitReady"
+	upgradeLogFluentdReady            = "FluentdReady"
+	upgradeLogClusterFlowAnnotation   = "harvesterhci.io/clusterFlow"
+	upgradeLogClusterOutputAnnotation = "harvesterhci.io/clusterOutput"
+	upgradeLogClusterFlowReady        = "ClusterFlowReady"
+	upgradeLogClusterOutputReady      = "ClusterOutputReady"
+	upgradeLogStateAnnotation         = "harvesterhci.io/upgradeLogState"
+	upgradeLogStateCollecting         = "Collecting"
+	upgradeLogStateStopped            = "Stopped"
+	archiveNameAnnotation             = "harvesterhci.io/archiveName"
+)
+
+type handler struct {
+	ctx                 context.Context
+	namespace           string
+	clusterFlowClient   ctlloggingv1.ClusterFlowClient
+	clusterOutputClient ctlloggingv1.ClusterOutputClient
+	daemonSetClient     ctlappsv1.DaemonSetClient
+	daemonSetCache      ctlappsv1.DaemonSetCache
+	deploymentClient    ctlappsv1.DeploymentClient
+	jobClient           ctlbatchv1.JobClient
+	jobCache            ctlbatchv1.JobCache
+	loggingClient       ctlloggingv1.LoggingClient
+	pvcClient           ctlcorev1.PersistentVolumeClaimClient
+	statefulSetClient   ctlappsv1.StatefulSetClient
+	statefulSetCache    ctlappsv1.StatefulSetCache
+	upgradeClient       ctlharvesterv1.UpgradeClient
+	upgradeCache        ctlharvesterv1.UpgradeCache
+	upgradeLogClient    ctlharvesterv1.UpgradeLogClient
+	upgradeLogCache     ctlharvesterv1.UpgradeLogCache
+}
+
+func (h *handler) OnUpgradeLogChange(_ string, upgradeLog *harvesterv1.UpgradeLog) (*harvesterv1.UpgradeLog, error) {
+	if upgradeLog == nil {
+		return upgradeLog, nil
+	}
+
+	if harvesterv1.UpgradeLogReady.GetStatus(upgradeLog) == "" {
+		logrus.Infof("[%s] Initialize upgradeLog %s/%s", upgradeLogControllerName, upgradeLog.Namespace, upgradeLog.Name)
+
+		toUpdate := upgradeLog.DeepCopy()
+		harvesterv1.UpgradeLogReady.CreateUnknownIfNotExists(toUpdate)
+
+		logrus.Infof("[%s] Deploy logging-operator", upgradeLogControllerName)
+		harvesterv1.OperatorDeployed.CreateUnknownIfNotExists(toUpdate)
+
+		// NOTE: As of v1.1.1, the logging-operator is by default deployed (rancher-logging), so we set the condition to true directly.
+		setOperatorDeployedCondition(toUpdate, corev1.ConditionTrue, "", "")
+		logrus.Infof("[%s] logging-operator deployed", upgradeLogControllerName)
+
+		return h.upgradeLogClient.Update(toUpdate)
+	}
+
+	if harvesterv1.OperatorDeployed.IsTrue(upgradeLog) && harvesterv1.InfraScaffolded.GetStatus(upgradeLog) == "" {
+		logrus.Infof("[%s] Start to scaffold the logging infrastructure for upgrade procedure", upgradeLogControllerName)
+
+		toUpdate := upgradeLog.DeepCopy()
+
+		if _, err := h.pvcClient.Create(preparePvc(upgradeLog)); err != nil && !apierrors.IsAlreadyExists(err) {
+			return nil, err
+		}
+		if _, err := h.loggingClient.Create(prepareLogging(upgradeLog)); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
Doesn't work in air-gapped env. Got `ImagePullBackOff` on these pods:
```
│ harvester-system           hvst-upgrade-hkvvl-upgradelog-infra-fluentbit-68mrp              ●  0/1          0 ErrImagePullΔ
│ harvester-system           hvst-upgrade-hkvvl-upgradelog-infra-fluentbit-dl6n6              ●  0/1          0 ImagePullBackOff
│ harvester-system           hvst-upgrade-hkvvl-upgradelog-infra-fluentbit-dnd9b              ●  0/1          0 ImagePullBackOff
│ harvester-system           hvst-upgrade-hkvvl-upgradelog-infra-fluentbit-jfpfm              ●  0/1          0 ImagePullBackOff
│ harvester-system           hvst-upgrade-hkvvl-upgradelog-infra-fluentd-0                    ●  0/2          0 ImagePullBackOff
```

The missing images are:
```
fluent/fluent-bit:1.9.3
jimmidyson/configmap-reload:v0.4.0
ghcr.io/banzaicloud/fluentd:v1.14.6-alpine-5
```

Need to figure out a way to determine what's available on the system.

Reply:
Change to rancher mirrored ones, which are already bundled in the ISO image. Please check it again, thanks!

Reply:
Yes, rancher mirroed ones are packed in ISO

https://github.com/rancher/charts/blob/dev-v2.7/charts/rancher-logging/100.1.1%2Bup3.17.3/values.yaml#L158